lib/libzpool/coroutine.c

```diff
@@ -0,0 +1,730 @@
+#include "atomic.h"
+#include "coroutine_impl.h"
+#include "sys/list.h"
+#include "sys/stdtypes.h"
+#include "sys/zfs_context.h"
+#include "umem.h"
+#include <asm-generic/errno-base.h>
+#include <asm-generic/errno.h>
+#include <bits/types/struct_timespec.h>
+#include <libintl.h>
+#include <pthread.h>
+#include <sched.h>
+#include <stdlib.h>
+#include <sys/ucontext.h>
+#include <time.h>
+#include <ucontext.h>
+#include <coroutine.h>
+#include <timer_thread.h>
+#include <stddef.h>
+#include <libuzfs.h>
+
+#define	MAX_COROUTINE_POOL_SIZE	100
+static __thread uzfs_coroutine_t *thread_local_coroutine = NULL;
+static __thread int cur_coroutine_pool_size = 0;
+
+static timer_thread_t timer_thread;
+static __thread uzfs_coroutine_t *coroutine_pool_head = NULL;
+static uint32_t cur_key_idx = 0;
+
+static inline void
+wakeup_coroutine(uzfs_coroutine_t *coroutine)
+{
+	VERIFY3U(atomic_cas_32(&coroutine->co_state, COROUTINE_PENDING,
+	    COROUTINE_RUNNABLE), ==, COROUTINE_PENDING);
+	coroutine->wake(coroutine->arg);
+}
+
+void
+cutex_init(cutex_t *cutex)
```

Comment:
iiuc, cutex is only for internal use, so it's better to add "static" for the cutex related functions.